test/test_log_warning.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import logging
+import sys
+import warnings
+from typing import Any, Callable, Type
+from unittest.mock import ANY, Mock
+
+import pytest
+
+from ops.log import setup_root_logging
+from ops.model import _ModelBackend
+
+
+def deprecated(msg: str, *, category: Type[Warning] = DeprecationWarning, stacklevel: int = 1):
+    def decorator(f: Callable[..., Any]) -> Callable[..., Any]:
+        @functools.wraps(f)
+        def inner(*args: Any, **kwargs: Any):
+            warnings.warn(category(msg), stacklevel=2)
+            return f(*args, **kwargs)
+
+        return inner
+
+    return decorator
+
+
+if sys.version_info >= (3, 13):
+    from warnings import deprecated  # type: ignore
```

Comment:
This feels unPythonic to me, but I'd be interested in hearing counter-arguments.

In my experience, this is typically done like:

```python
try:
    from warnings import deprecated
except ImportError:
    # Handle Python < 3.13.
    # backport goes here
```

Even when it's not an import, e.g.:

```python
try:
    True, False
except NameError:
    True, False = 1, 0  # Handle Python < 2.2.2
```

Reply:
try/except makes `pyright` evaluate the type of `deprecated`, which it cannot until we've got a Py 3.13 runtime (as in all devs + ci), and I couldn't find a good way to suppress this type interference, other than to hide the statement behind an `if` clause.

Reply:
Hmm, interesting. My thinking would be Pythonic+type:ignore > unPythonic+check-version. However, I could well be biased by my general distaste for typing in Python :smile:. I'm curious what the consensus here would be.

Reply:
That's a discussion I'd like to have, because I could as well adopt the output of it when deprecating the framework methods that rely on AllWatcher.

Reply:
P.S. I'd love to use:

```
try:
    from warnings import deprecated
except ImportError:
    from .utils.backport import deprecated
```

I just couldn't get the type checker happy.
The stdlib decorator is already included in typeshed, so maybe there's soon a way to resolve the types and then there's no need for discussion 🎉